src/middlewared/middlewared/plugins/pool_/dataset.py

```diff
@@ -531,23 +531,38 @@ async def do_create(self, data):
                     'format-options': {'canonicalize': True, 'ensure_builtins': True},
                 }))[0]['acl']
         elif data['share_type'] == 'APPS':
+            must_add_apps = True
             if parent_st['acl'] and parent_st['acltype'] == 'NFS4':
                 acl_to_set = await self.middleware.call('filesystem.get_inherited_acl', {
                     'path': os.path.join('/mnt', parent_name),
                 })
+
+                # The inherited ACL may already contain an entry granting MODIFY permissions.
+                # if it does, then we can skip adding the apps entry.
+                for entry in acl_to_set:
+                    if entry['id'] == 568 and entry['tag'] == 'USER' and entry['type'] == 'ALLOW':
+                        if entry['flags']['FILE_INHERIT'] and entry['flags']['DIRECTORY_INHERIT']:
+                            if all([entry['perms'][x] for x in [
+                                'READ_DATA', 'WRITE_DATA', 'DELETE', 'DELETE_CHILD', 'READ_ACL',
+                                'APPEND_DATA', 'READ_NAMED_ATTRS', 'WRITE_NAMED_ATTRS',
+                                'READ_ATTRIBUTES', 'WRITE_ATTRIBUTES'
+                            ]]):
```

Comment:
```suggestion
                            if all(
                                entry['perms'][role]
                                for role in (
                                    'READ_DATA', 'WRITE_DATA', 'DELETE', 'DELETE_CHILD', 'READ_ACL', 'APPEND_DATA',
                                    'READ_NAMED_ATTRS', 'WRITE_NAMED_ATTRS', 'READ_ATTRIBUTES', 'WRITE_ATTRIBUTES'
                                )
                            ):
```
A micro-optimization (short circuits instead of building and evaluating a whole list)